frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx

```diff
@@ -0,0 +1,108 @@
+import React from "react";
+import { Route } from "react-router";
+import { screen } from "@testing-library/react";
+import userEvent from "@testing-library/user-event";
+import fetchMock from "fetch-mock";
+import { renderWithProviders, waitForElementToBeRemoved } from "__support__/ui";
+import DashboardApp from "metabase/dashboard/containers/DashboardApp";
+import { createMockDashboard, createMockUser } from "metabase-types/api/mocks";
+import {
+  setupCollectionItemsEndpoint,
+  setupCollectionsEndpoints,
+  setupDashboardEndpoints,
+} from "__support__/server-mocks";
+import { createMockDashboardState } from "metabase-types/store/mocks";
+import { BEFORE_UNLOAD_UNSAVED_MESSAGE } from "metabase/dashboard/constants";
+
+const mockDashboard = createMockDashboard({
+  id: 1,
+  name: "Example",
+  parameters: [
+    { id: "1", name: "Example", type: "type/Text", slug: "my_param" },
+  ],
+});
+
+async function setup(user = createMockUser()) {
+  setupDashboardEndpoints(createMockDashboard(mockDashboard));
+  setupCollectionsEndpoints([]);
+  setupCollectionItemsEndpoint();
+
+  fetchMock.get("path:/api/bookmark", []);
+
+  jest.spyOn(document, "querySelector").mockImplementation(() => ({
+    setAttribute: jest.fn(),
+    classList: {
+      remove: jest.fn(),
+    },
+  }));
+
+  jest
+    .spyOn(document, "getElementsByTagName")
+    .mockImplementation(() => [
+      { addEventListener: jest.fn(), removeEventListener: jest.fn() },
+    ]);
+
+  const { mockEvent, events } = createMockEventListener();
+
+  renderWithProviders(
+    <Route path="/dashboard/:slug" component={DashboardApp} />,
+    {
+      initialRoute: "/dashboard/1",
+      currentUser: user,
+      withRouter: true,
+      storeInitialState: {
+        dashboard: createMockDashboardState(),
+      },
+    },
+  );
+
+  await waitForElementToBeRemoved(() =>
+    screen.queryAllByTestId("loading-spinner"),
+  );
+
+  return { mockEvent, events };
+}
+
+function createMockEventListener() {
+  const events = {};
+  const mockEvent = {
+    preventDefault: jest.fn(),
+    returnValue: undefined,
+  };
+
+  window.addEventListener = jest.fn((event, callback) => {
```

Comment:
It seems that we don't revert this change. Maybe `jest.spyOn(window, "addEventListener")` and then `jest.restoreAllMocks();` in `afterEach` would do the trick?

Reply:
yep, makes sense and will help with consistency with the other code